packages/roosterjs-plugin-react/lib/ReactPlugin.tsx

```diff
@@ -0,0 +1,314 @@
+     */
+    inEditorMountRoot: HTMLElement;
+    /**
+     * React root where the actual react component has been mounted
+     */
+    offDocumentReactRoot: HTMLElement;
+    instanceId: string;
+    updateMountPointCallback?: (element: HTMLElement) => void;
+}
+
+const legalChangeSources = new Set();
+legalChangeSources.add(ChangeSource.SetContent);
+legalChangeSources.add(ChangeSource.Paste);
+legalChangeSources.add(ChangeSource.Drop);
+
+export default class ReactComponentPlugin implements EditorPlugin {
+    private editor: Editor | undefined;
+    private nextId: number = 0;
+    private elementToInstances: Map<HTMLElement, MountedComponentInstance> = new Map();
+    private idToInstances: Map<string, MountedComponentInstance> = new Map();
+    private componentId: string;
+    private componentType: React.ComponentType<ReactPluginComponentProps>;
+
+    constructor(componentId: string, componentType: React.ComponentType<ReactPluginComponentProps>) {
+        this.componentId = componentId;
+        this.componentType = componentType;
+    }
+
+    /**
+     * Get a friendly name
+     */
+    getName() {
+        return "React Component";
+    }
+
+    /**
+     * Initialize this plugin. This should only be called from Editor
+     * @param editor Editor instance
+     */
+    public initialize(editor: Editor) {
+        this.editor = editor;
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose() {
+        this.editor = undefined;
+        for (let mountedInstance of [...this.elementToInstances.values()]) {
+            this.unmountInstance(mountedInstance);
+        }
+    }
+
+    /**
+     * Check if the plugin should handle the given event exclusively.
+     * Handle an event exclusively means other plugin will not receive this event in
+     * onPluginEvent method.
+     * If two plugins will return true in willHandleEventExclusively() for the same event,
+     * the final result depends on the order of the plugins are added into editor
+     * @param _event The event to check
+     */
+    public willHandleEventExclusively(_event: PluginEvent) {
+        return false;
+    }
+
+    /**
+     * Handle events triggered from editor
+     * @param event PluginEvent object
+     */
+    public onPluginEvent(event: PluginEvent) {
+        if (this.isTriggeringEvent(event)) {
+            this.handleChangeInEditor(event);
+        } else if (event.eventType === PluginEventType.ExtractContent) {
+            const evt = event as ExtractContentEvent;
+            evt.content = removeReactTagsOnExtractContent(evt.content);
+        }
+    }
+
+    private handleChangeInEditor(event: PluginEvent) {
+        if (!this.editor) {
+            return;
+        }
+
+        // Undoes and other major changes to document content fire this type of event.
+        //
+        // Scan for new elements.
+        const foundComponentRoots = this.editor.queryElements(`[${REACT_COMPONENT_DATA_KEY}="${this.componentId}"]`);
```

Comment:
Does this mean multiple react components share the same component id?

Reply:
the `componentId` here, is an identifier of the react component type (e.g. is shared by all components with the same component constructor, or component function).

This is so you can host multiple types of components in the rooster js editor.

I'll rename `componentTypeId` in the class and document it